### PR TITLE
When $db connection is down, generate fewer logs

### LINKED ISSUE #5722: $db, should it just `die()` when the database connection fails?
Having an issue with one of my sites' servers, where the initial problem is reported as a PHP Warning:
```
[02-May-2023 13:14:55 America/Chicago] Request URI: /search-results?search_in_description=1&keyword=altemp&inc_subcat=0&page=1&sort=2d, IP address: xx.xx.xx.xx
#1  mysqli_connect() called at [/home/mysite/public_html/includes/classes/db/mysql/query_factory.php:64]
#2  queryFactory->connect() called at [/home/mysite/public_html/includes/init_includes/init_database.php:24]
#3  require(/home/mysite/public_html/includes/init_includes/init_database.php) called at [/home/mysite/public_html/includes/autoload_func.php:48]
#4  require(/home/mysite/public_html/includes/autoload_func.php) called at [/home/mysite/public_html/includes/application_top.php:170]
#5  require(/home/mysite/public_html/includes/application_top.php) called at [/home/mysite/public_html/index.php:26]
--> PHP Warning: mysqli_connect(): (HY000/2002): No such file or directory in /home/mysite/public_html/includes/classes/db/mysql/query_factory.php on line 64.
```
This then leads to 9Kb+ log files created on each site access (and it's a busy site) for as long as the database issue's present.

If the `mysqli_connect` isn't successful, shouldn't the page-load just 'die'?

### PR DESCRIPTION
**Pull request by drbyte**
Fixes #5722

- suppresses E_WARNING which triggers logging, on simpleConnect()
- dies sooner on connect()

### CODE REVIEW

**Comment by drbyte** [comment]
I've tested this on zc_install, where it now no longer generates 20+ logged failures.

Will need testing on the scenarios @lat9 was mentioning in the original issue.

File: includes/classes/db/mysql/query_factory.php [hunk 1]
```diff
@@ -72,8 +72,15 @@ public function connect(string $db_host, string $db_user, string $db_password, s
         mysqli_report(MYSQLI_REPORT_OFF);
 
         $connectionRetry = 10;
-        while (!isset($this->link) || ($this->link == false && $connectionRetry != 0)) {
+        while (!isset($this->link) || ($this->link == false && $connectionRetry > 0)) {
             $this->link = mysqli_connect($db_host, $db_user, $db_password, $db_name, (defined('DB_PORT') ? DB_PORT : null), (defined('DB_SOCKET') ? DB_SOCKET : null));
+
+            // mariadb connection down or incorrect
+            if (mysqli_connect_errno() == 2002) {
```
Comment (lat9) [inline includes/classes/db/mysql/query_factory.php:79]: Use `===`.

File: includes/classes/db/mysql/query_factory.php [hunk 2]
```diff
@@ -72,8 +72,15 @@ public function connect(string $db_host, string $db_user, string $db_password, s
...
+                if ($connectionRetry > 1) $connectionRetry = 1; // if service is down, try only one more time
```
Comment (lat9) [inline includes/classes/db/mysql/query_factory.php:80]: PSR, no single-line conditionals.

File: includes/classes/db/mysql/query_factory.php [hunk 3]
```diff
@@ -127,7 +134,11 @@ public function simpleConnect($db_host, $db_user, $db_password, $db_name): bool
         $this->user = $db_user;
         $this->host = $db_host;
         $this->password = $db_password;
+
+        // temporarily suppress E_WARNING in case of connection failure
+        $error_level = error_reporting(E_ERROR | E_PARSE);
         $this->link = mysqli_connect($db_host, $db_user, $db_password, $db_name, (defined('DB_PORT') ? DB_PORT : null), (defined('DB_SOCKET') ? DB_SOCKET : null));
+        error_reporting($error_level);
```
Comment (lat9) [inline includes/classes/db/mysql/query_factory.php:141]: Nice!

File: includes/classes/db/mysql/query_factory.php [hunk 4]
```diff
@@ -645,17 +658,37 @@ protected function show_error()
...
         }
-        if ($this->error_number == 0 && $this->error_text == DB_ERROR_NOT_CONNECTED && file_exists(FILENAME_DATABASE_TEMPORARILY_DOWN)) {
-            include(FILENAME_DATABASE_TEMPORARILY_DOWN);
+
+        if (!defined('FILENAME_DATABASE_TEMPORARILY_DOWN')) {
+            define('FILENAME_DATABASE_TEMPORARILY_DOWN', DIR_FS_CATALOG . '/nddbc.html');
+        }
+        if (file_exists(FILENAME_DATABASE_TEMPORARILY_DOWN)) {
+            if (($this->error_number == 0 && $this->error_text == DB_ERROR_NOT_CONNECTED)
+                || $this->error_number == 2002)
```
Comment (lat9) [inline includes/classes/db/mysql/query_factory.php:667]: Use `===`

File: includes/classes/db/mysql/query_factory.php [hunk 5]
```diff
@@ -645,17 +658,37 @@ protected function show_error()
...
+            {
+                include(FILENAME_DATABASE_TEMPORARILY_DOWN);
+            }
+        }
+
+        // suppress backtrace for MariaDB connection errors: not logging these because they usually come hundreds at a time
+        if ($this->error_number == 2002 || defined('DIR_FS_INSTALL')) {
```
Comment (lat9) [inline includes/classes/db/mysql/query_factory.php:674]: Use `===`

File: includes/classes/db/mysql/query_factory.php [hunk 6]
```diff
@@ -645,17 +658,37 @@ protected function show_error()
...
+
+        // display error details if appropriate
         echo '<div class="systemError">';
         if (defined('STRICT_ERROR_REPORTING') && STRICT_ERROR_REPORTING == true) {
             echo $this->error_number . ' ' . $this->error_text;
             echo '<br>in:<br>[' . (strstr($this->zf_sql, 'db_cache') ? 'db_cache table' : $this->zf_sql) . ']<br>';
         } else {
             echo 'WARNING: An Error occurred, please let us know!';
         }
+        if (defined('IS_ADMIN_FLAG') && IS_ADMIN_FLAG == true) {
```
Comment (lat9) [inline includes/classes/db/mysql/query_factory.php:686]: Use `===`

**Review by lat9 — CHANGES_REQUESTED** [review]
Me, as usual, being a picky-arse.

I'll grab this and see how it performs on some of my managed sites that are getting hammered by spiders.

**Comment by scottcwilson** [comment]
Thanks, this will help reduce the quantity of logs.

**Comment by lat9** [comment]
Just tested before/after for a shopping-cart page-load where the MySQL server is shut down at the start of the page load.

The current processing takes about 15 seconds and 6Kb of duplicated entries in a log to report what this update does in about 5 seconds with a single-entry log!

**Comment by drbyte** [comment]
> ... what this update does in about 5 seconds with a single-entry log!

Yay!

Any idea what the 5-seconds is about? (vs just milliseconds)

**Comment by scottcwilson** [comment]
Great job, team!  Exciting progress.

**Comment by lat9** [comment]
> > ... what this update does in about 5 seconds with a single-entry log!
> 
> Yay!
> 
> Any idea what the 5-seconds is about? (vs just milliseconds)

That was just a "visual" note that it took _significantly_ longer with the current version than with your update.

**Comment by lat9** [comment]
Why are all my picky-arse comments shown as resolved when nothing's been changed?

File: includes/classes/db/mysql/query_factory.php (hunk 2 above)
Comment (drbyte) [reply includes/classes/db/mysql/query_factory.php]: Sigh. I really despise using 3-lines for a 1-line thing that's actually clearer to read when written as 1-line.
But, done.

File: includes/classes/db/mysql/query_factory.php (hunk 6 above)
Comment (drbyte) [reply includes/classes/db/mysql/query_factory.php:689]: That can be done separately. I didn't add/change this code in this PR, merely relocated it.

File: includes/classes/db/mysql/query_factory.php (hunk 1 above)
Comment (drbyte) [reply includes/classes/db/mysql/query_factory.php]: Rewritten.

File: includes/classes/db/mysql/query_factory.php (hunk 4 above)
Comment (drbyte) [reply includes/classes/db/mysql/query_factory.php]: Rewritten.

File: includes/classes/db/mysql/query_factory.php (hunk 5 above)
Comment (drbyte) [reply includes/classes/db/mysql/query_factory.php]: Rewritten.

**Comment by lat9** [comment]
> > > ... what this update does in about 5 seconds with a single-entry log!
> > 
> > 
> > Yay!
> > Any idea what the 5-seconds is about? (vs just milliseconds)
> 
> That was just a "visual" note that it took _significantly_ longer with the current version than with your update.

The reduced time from the previous implementation _could_ have something to do with the reduced volume of logging.

**Pull request merged**